packages/grafana-data/src/field/fieldDisplay.ts

```diff
@@ -65,10 +68,69 @@ function getTitleTemplate(title: string | undefined, stats: string[], data?: Dat
   if (fieldCount > 1 || !parts.length) {
     parts.push('${' + VAR_FIELD_NAME + '}');
   }
+  parts.push('${' + VAR_FIELD_LABELS + '}');
 
   return parts.join(' ');
 }
 
+/**
+ * Get an appropriate display title
+ */
+export function getFrameDisplayTitle(frame: DataFrame, index?: number) {
+  if (frame.name) {
+    return frame.name;
+  }
+
+  // Single field with tags
+  const valuesWithLabels = frame.fields.filter(f => f.labels !== undefined);
+  if (valuesWithLabels.length === 1) {
+    return formatLabels(valuesWithLabels[0].labels!);
+  }
+
+  // list all the
+  if (index === undefined) {
+    return frame.fields
+      .filter(f => f.type !== FieldType.time)
+      .map(f => getFieldDisplayTitle(f, frame))
+      .join(', ');
+  }
+
+  if (frame.refId) {
+    return `Series (${frame.refId})`;
+  }
+
+  return `Series (${index})`;
+}
+
+/**
+ * Get an appropriate display title.  If the 'title' is set, use that
+ */
+export function getFieldDisplayTitle(field: Field, frame?: DataFrame) {
+  let title = field.config?.title;
+  if (title) {
+    return title; // Title is set and not a template
+  }
+
+  const id = getFieldId(field);
+  const seriesName = frame?.name;
+  if (seriesName && name !== seriesName) {
+    return seriesName + ' ' + name;
+  }
+  return id;
+}
```

Comment:
The biggest problem here is that we do not have access to other data frames here to dedupe and remove frame name when the frame name is all the same.

Currently, we have a big breaking behavior for Prometheus still. A query like `sum by(mode)(irate(node_cpu_seconds_total{job="node_exporter", instance=~".*", mode!="idle"}[5m])) > 0`  Returns no __name__ label (because of the sum). So with your updated code in Prometheus result_transformer the frame names are all the same (the query text). So the series names are all:

![Screenshot from 2020-04-30 09-37-57](https://user-images.githubusercontent.com/10999/80684688-55735180-8ac6-11ea-9651-15d15641f9b6.png)

So the frame name is all the same for all the series. It would be easy to fix the logic in the prometheus result transformer to not set frame name to options.query if there are any labels. So, in that case, we can have an empty frame name. Hm.. maybe that is correct. But then it would be impossible to separate two queries that return the same label sets. But maybe not a common/realistic scenario.

Reply:
So to add, that query today (in 6.7) would result in a legend with only {mode="value"} ie only the label key-value pair.

Reply:
I changed the result_transformer in prometheus to set name to undefined if there are other labels (but no __name__) label. But then I get

Value prefix, and here is the problem of not knowing the names of the fields in the other data frames in order to have smart title info removal for duplicated (same) names  :) 
![Screenshot from 2020-04-30 09-47-05](https://user-images.githubusercontent.com/10999/80685541-a768a700-8ac7-11ea-9885-4305c9899034.png)